design/ingress-status-loadbalancer.md

```diff
@@ -0,0 +1,101 @@
+
+### Dynamic load balancer address discovery
+
+The far more common case will be dynamic load balancer discovery.
+In this model Contour will track a service document (default to projectcontour/envoy, but configurable) and use those details as the list of ip addresses to write back to the ingress' status.
+
+The nginx controller goes to some lengths to try to dynamically figure out what it's status document as well as other side channels to discover the address if that fails.
+For this first implementation I propose we stick to the discovery methods outlined above.
+
+### Status updater
+
+The latter, writing the status to valid ingress objects, waits to be elected the leader then updates the status document on all of the ingress objects within Contour's class.
+
+As new ingress objects arrive they will need to be updated.
+
+Note: there appears to be no requirement that if Contour is shutting down we delete the status field on the ingress on the way out.
+Initially this feels wrong but consider the scenarios:
+
+- Leadership has deposed the current Contour causing it to shut down; the new leader will be responsible for updating status. Best that during this transition observers of the Ingress document do not see status disappear and reappear quickly.
+- Contour is shut down for a long period of time. Yes, the status information will be incorrect, but there is no correct value, no ingress controller is serving that vhost.
+- Contour is being replaced with something else, then the responsibility to _overwrite_ the status information falls to Contour's replacement.
+
+## Detailed Design
+
+I think the design should involve two goroutines added to cmd/contour's workgroup.
+The first will be responsible for load balancer discovery, the second will be responsible for updating status documents.
+
+### Discovery worker
+
+The discovery worker should look something like this pseudocode:
+
+1. worker starts up, blocks on leadership election
+2. when elected leader it checks the mode
+2a. if discovery is hard coded in the config file it transmits these addresses over a channel to the other worker, then goes to sleep until deposed and/or shutdown.
+2b. if discovery is dynamic it opens a watcher on the specified (defaults, config file, cli flags, in that order) service document and each time it changes sends the address details to the other worker until deposed or shutdown. 
+
+Open question: if the service document goes away, should we transmit an empty set of addresses to the status worker to indicate it should remove the status.loadbalancer details?
+
+The send from the discovery worker to the status worker will be over an unbuffered channel so the discovery worker cannot send until the status worker has been elected leader and is ready to receive ip details.
+Care should be taken to ensure the discovery worker can respond to the group's shutdown signal.
+
+### Status worker
+
+The status worker should look something like this pseudocode:
```

Comment:
Is there any relationship between the load balancer status and the Ingress validity? We should call that out here.